system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1302,355 @@ handle_run_triggers (FlatpakSystemHelper   *object,
   return TRUE;
 }
 
+static gboolean
+get_connection_uid (GDBusMethodInvocation *invocation, uid_t *out_uid, GError **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  g_autoptr(GVariant) dict = NULL;
+  g_autoptr(GVariant) credentials = NULL;
+
+  credentials = g_dbus_connection_call_sync (connection,
+                                             "org.freedesktop.DBus",
+                                             "/org/freedesktop/DBus",
+                                             "org.freedesktop.DBus",
+                                             "GetConnectionCredentials",
+                                             g_variant_new ("(s)", sender),
+                                             G_VARIANT_TYPE ("(a{sv})"), G_DBUS_CALL_FLAGS_NONE,
+                                             G_MAXINT, NULL, error);
+  if (credentials == NULL)
+    return FALSE;
+
+  dict = g_variant_get_child_value (credentials, 0);
+
+   if (!g_variant_lookup (dict, "UnixUserID", "u", out_uid))
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                   "Failed to query UnixUserID for the bus name: %s", sender);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+check_revokefs_requirements (GError **error)
+{
+  g_autofree gchar *revokefs_fuse_path = NULL;
+  struct passwd *passwd;
+
+  revokefs_fuse_path = g_find_program_in_path ("revokefs-fuse");
+  if (revokefs_fuse_path == NULL)
+    {
+      g_set_error (error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED,
+                   "Can't find revoke-fuse on the system.");
+      return FALSE;
+    }
+
+  passwd = getpwnam ("flatpak");
```

Comment:
It's a bit sad to call `getpwnam()` 3 times for each call to `GetRevokefsFD`.

I'm also reasonably sure that it's not safe to call `getpwnam` in multi-threaded programs. These functions which return a pointer that you don't have to free are normally implemented by using a `static` variable defined in the function; subsequent calls re-use the same block of memory.

You probably want to use `getpwnam_r`, which stores the result into a block of memory that the caller allocates. It's one of these great APIs where you have to retry with ever-bigger buffers until it succeeds; you could either just pick some large-ish buffer size (as the example in `man getpwnam_r` shows), or use a loop to retry with ever-bigger buffers if it fails with `ERANGE` (as in glib/gutils.c g_get_user_database_entry()).

You could do this once, on entry to ongoing_pull_new (or indeed in this function) and use the returned value in the two places it's needed:

1. setting the owner on the underlying directory
2. `setuid`/`setgid` in the child_setup function (pass the allocated buffer as the user_data)

This would also avoid the (very unlikely) case that getpwnam() succeeds here but fails in the child_setup function.